File: src/DroidSolutions.Oss.JobService/Worker/JobWorkerBase.cs
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;
using System.Diagnostics.Metrics;

using DroidSolutions.Oss.JobService.Worker.Settings;

using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;

using NanoidDotNet;

using Reinforced.Typings.Attributes;

namespace DroidSolutions.Oss.JobService.Worker;

/// <summary>
/// A base worker for recurrent jobs managed in a database.
/// </summary>
/// <typeparam name="TParams">The type of the parametersa job can have.</typeparam>
/// <typeparam name="TResult">The type of the result a job can have.</typeparam>
[TsInterface]
public abstract class JobWorkerBase<TParams, TResult> : BackgroundService, IJobWorker
  where TParams : class?
  where TResult : class?
{
  /// <summary>
  /// Gets the meter for this worker.
  /// </summary>
  protected static readonly Meter WorkerMeter = new("droidsolutions.oss.jobworker");
  private static readonly Counter<int> ExecutedJobsCounter = WorkerMeter.CreateCounter<int>(
    name: "executed_jobs",
    unit: "{jobs}",
    description: "Counter for executed jobs");

  private static readonly Histogram<double> JobProcessingTime = WorkerMeter.CreateHistogram<double>(
    name: "job_processing_time",
    unit: "ms",
    description: "Histogram for job processing time");

  private readonly IOptionsMonitor<JobWorkerSettings> _workerSettings;
  private readonly IServiceProvider _serviceProvider;
  private readonly ILogger _logger;
  private CancellationToken _stopToken;
  private IJobRepository<TParams, TResult>? _jobRepository;
  private IJob<TParams, TResult>? _currentJob;
  private int _executedJobs;
  private long _lastJobDurationMs;
  private DateTime? _lastJobFinishedAt;
  private DateTime? _lastDeletedAt;

  /// <summary>
  /// Initializes a new instance of the <see cref="JobWorkerBase{TParams, TResult}"/> class.
  /// </summary>
  /// <param name="workerSettings">An instance of an monitor for worker settings.</param>
  /// <param name="serviceProvider">An instance of the service provider.</param>
  /// <param name="logger">An instance of the logger.</param>
  protected JobWorkerBase(
    IOptionsMonitor<JobWorkerSettings> workerSettings,
    IServiceProvider serviceProvider,
    ILogger<JobWorkerBase<TParams, TResult>> logger)
  {
    _workerSettings = workerSettings;
    _serviceProvider = serviceProvider;
    _logger = logger;

    RunnerName = nameof(JobWorkerBase<TParams, TResult>);
  }

  /// <summary>
  /// Gets the name of the current runner.
  /// </summary>
  public string RunnerName { get; private set; }

  /// <summary>
  /// Gets the last time the job execution loop started.
  /// </summary>
  [TsIgnore]
  protected DateTime? LastJobExecutionStart { get; private set; }

  /// <summary>
  /// Gets the last time the job execution loop finished.
  /// </summary>
  [TsIgnore]
  protected DateTime? LastJobExecutionStop { get; private set; }

  /// <summary>
  /// Exports collected metrics.
  /// </summary>
  /// <returns>An object with <c>ExecutedJobs</c> and <c>LastJobDurationMs</c>.</returns>
  public JobWorkerMetrics GetMetrics()
  {
    return new JobWorkerMetrics(_workerSettings.CurrentValue.JobType)
    {
      ExecutedJobs = _executedJobs,
      LastJobDurationMs = _lastJobDurationMs,
      JobIntervallSeconds = (int?)_workerSettings.CurrentValue.AddNextJobAfter?.TotalSeconds,
      LastJobFinishedAt = _lastJobFinishedAt,
    };
  }

  /// <summary>
  /// Should return the name of the runner. A random string is attached to it so it will be unique for each instance.
  /// </summary>
  /// <returns>A string representing the name of the runner.</returns>
  protected abstract string GetRunnerName();

  /// <summary>
  /// Returns the parameters that are used if an initial job is created.
  /// </summary>
  /// <remarks>
  /// <p>If an initial job is created the result of this function is used as the parameters of the job. The method can
  /// be overwritten to provide actual parameters, otherwise null is used.</p>
  /// <p>Only applies if an inital job is created.</p>
  /// <p>If configured the initial job will set a new job at the end using the same parameters.</p>
  /// </remarks>
  /// <returns>The parameters for the initial job.</returns>
  [TsFunction(Type = "TParams | undefined")]
  protected virtual TParams? GetInitialJobParameters()
  {
    return null;
  }

  /// <summary>
  /// Hook that is called when a job run starts.
  /// </summary>
  protected virtual void PreJobRunHook()
  {
  }

  /// <summary>
  /// Hook that is called when a job run ends.
  /// </summary>
  protected virtual void PostJobRunHook()
  {
  }

  /// <summary>
  /// Processes the job and returns the result if any.
  /// </summary>
  /// <param name="job">The job.</param>
  /// <param name="serviceScope">A service scope for injecting additional services.</param>
  /// <param name="cancellationToken">A token to cancel the operation.</param>
  /// <returns>The result of the job, if any.</returns>
  protected abstract Task<TResult?> ProcessJobAsync(
    IJob<TParams, TResult> job,
    [TsIgnore] IServiceScope serviceScope,
    [TsParameter(Type = "AbortSignal")] CancellationToken cancellationToken);

  /// <inheritdoc/>
  protected override async Task ExecuteAsync(
    [TsParameter(Type = "AbortSignal")] CancellationToken stoppingToken)
  {
    _stopToken = stoppingToken;

    var runnerId = Nanoid.Generate("0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ", 7);
    RunnerName = $"{GetRunnerName()}-{runnerId}";

    _logger.LogInformation(
      "Delaying runner {Runner} start for {Seconds} seconds.",
      RunnerName,
      _workerSettings.CurrentValue.InitialDelaySeconds);
    await Task.Delay(_workerSettings.CurrentValue.InitialDelaySeconds * 1000, stoppingToken);

    _logger.LogInformation("Starting runner {Runner}", RunnerName);
    var isFirstRun = true;

    while (!stoppingToken.IsCancellationRequested)
    {
      Stopwatch stopwatch = new();
      stopwatch.Start();
      LastJobExecutionStart = DateTime.UtcNow;
      var executed = false;

      try
      {
        PreJobRunHook();
        JobWorkerSettings settings = _workerSettings.CurrentValue;
        executed = await HandleJobRunAsync(settings, stoppingToken, isFirstRun);

        PostJobRunHook();
        isFirstRun = false;
      }
      catch (OperationCanceledException) when (stoppingToken.IsCancellationRequested)
      {
        _logger.LogInformation("Cancellation requested, worker {Runner} stopped.", RunnerName);
        break;
      }
      catch (Exception ex)
      {
        _logger.LogError(ex, "Stopping worker {Runner} due to an unhandled error.", RunnerName);
        throw;
      }
      finally
      {
        stopwatch.Stop();
        if (executed)
        {
          _lastJobFinishedAt = DateTime.UtcNow;
          _lastJobDurationMs = stopwatch.ElapsedMilliseconds;
          JobProcessingTime.Record(
            _lastJobDurationMs,
            new KeyValuePair<string, object?>("type", _workerSettings.CurrentValue.JobType));
        }

        LastJobExecutionStop = DateTime.UtcNow;
      }

      try
      {
        stoppingToken.ThrowIfCancellationRequested();
        JobWorkerSettings settings = _workerSettings.CurrentValue;

        // Wait before checking again for available jobs
        await Task.Delay(settings.JobPollingIntervalSeconds * 1000, stoppingToken);
      }
      catch (OperationCanceledException) when (stoppingToken.IsCancellationRequested)
      {
        _logger.LogInformation("Cancellation requested, worker {Runner} stopped.", RunnerName);
        break;
      }
    }
  }

  /// <summary>
  /// Updates the total items of the current job.
  /// </summary>
  /// <param name="items">The amount of items the job must process.</param>
  /// <returns>A task indicating when the operation is complete.</returns>
  /// <exception cref="InvalidOperationException">When no current job or repository is set.</exception>
  protected async Task SetTotalItemsAsync(int items)
  {
    CheckJobAndRepo();

    await _jobRepository.SetTotalItemsAsync(_currentJob, items, _stopToken);
  }

  /// <summary>
  /// Updates the given amount to successful items on the current job.
  /// </summary>
  /// <param name="amount">The amount to add, 1 if nothing given.</param>
  /// <returns>A taks indicating when the operation is complete.</returns>
  /// <exception cref="InvalidOperationException">When no current job or repository is set.</exception>
  protected async Task AddProgressAsync(int amount = 1)
  {
    CheckJobAndRepo();

    await _jobRepository.AddProgressAsync(_currentJob, amount, false, _stopToken);
  }

  /// <summary>
  /// Updates the given amount to successful items on the current job.
  /// </summary>
  /// <param name="amount">The amount to add, 1 if nothing given.</param>
  /// <returns>A taks indicating when the operation is complete.</returns>
  /// <exception cref="InvalidOperationException">When no current job or repository is set.</exception>
  protected async Task AddFailedProgressAsync(int amount = 1)
  {
    CheckJobAndRepo();

    await _jobRepository.AddProgressAsync(_currentJob, amount, true, _stopToken);
  }

  /// <summary>
  /// Retrieves a time that marks the span between finishing the current job and the due date of the next job. If null,
  /// no new job will be added.
  /// </summary>
  /// <param name="settings">An instance of the current job settings.</param>
  /// <param name="result">The result of the current job.</param>
  /// <returns>The timespan from now until the next job is due.</returns>]
  [TsFunction(Type = "TimeSpan | undefined")]
  protected virtual TimeSpan? AddNextJobIn(JobWorkerSettings settings, TResult? result)
  {
    return settings.AddNextJobAfter;
  }

  /// <summary>
  /// Handles a single run by checking for a new job and invoke the processing method if any.
  /// </summary>
  /// <param name="settings">An instance of the current job settings.</param>
  /// <param name="cancellationToken">A token to cancel the run.</param>
  /// <param name="firstRun">A value indicating whether this is the first run of this worker.</param>
  /// <returns>A task indicating when the run is finished.</returns>
  /// <exception cref="InvalidOperationException">
  /// When an exception is thrown during checking for a new job.
  /// </exception>
  [TsIgnore]
  private async Task<bool> HandleJobRunAsync(
    JobWorkerSettings settings,
    CancellationToken cancellationToken,
    bool firstRun = false)
  {
    var executed = false;
    using IServiceScope serviceScope = _serviceProvider.CreateScope();
    _jobRepository = serviceScope.ServiceProvider.GetRequiredService<IJobRepository<TParams, TResult>>();

    _currentJob = await GetJobAsync(_jobRepository, settings, cancellationToken);

    if (_currentJob != null)
    {
      try
      {
        _currentJob.Result = await ProcessJobAsync(_currentJob, serviceScope, cancellationToken);
        TimeSpan? nextJobIn = AddNextJobIn(settings, _currentJob.Result);
        await _jobRepository.FinishJobAsync(_currentJob, nextJobIn, cancellationToken);

        _executedJobs++;
        ExecutedJobsCounter.Add(1, new KeyValuePair<string, object?>("type", _currentJob.Type));
        executed = true;
      }
      catch (Exception ex)
      {
        _logger.LogWarning(
          ex,
          "Failed to process job {JobId}, job is beeing resetted. Message: {Message}",
          _currentJob.Id,
          ex.Message);

        // Since operation was cancelled (probably cause app is shutting down) no sense in forwarding our token
        await _jobRepository.ResetJobAsync(_currentJob, default);

        if (ex is OperationCanceledException && cancellationToken.IsCancellationRequested)
        {
          throw;
        }
      }
    }
    else if (firstRun && settings.AddInitialJob)
    {
      await AddInitialJob(settings, cancellationToken);
    }

    _currentJob = null;

    if (settings.DeleteJobsOlderThan.HasValue)
    {
      try
      {
        await DeleteOldJobs(settings.DeleteJobsOlderThan.Value, settings.JobType, cancellationToken);
      }
      catch (Exception ex)
      {
        _logger.LogError(ex, "Error deleting old jobs: {Message}", ex.Message);
      }
    }

    _jobRepository = null;

    return executed;
  }

  /// <summary>
  /// Checks if there is a job available and starts the one with the oldest due date that is applicable.
  /// </summary>
  /// <param name="repo">An instance of the job repository.</param>
  /// <param name="settings">An instance of the settings for the runner.</param>
  /// <param name="cancellationToken">A token to cancel the operation.</param>
  /// <returns>The first applicable job or null if no job is applicable.</returns>
  [TsIgnore]
  private async Task<IJob<TParams, TResult>?> GetJobAsync(
  IJobRepository<TParams, TResult> repo,
  JobWorkerSettings settings,
  CancellationToken cancellationToken)
  {
    try
    {
      return await repo.GetAndStartFirstPendingJobAsync(settings.JobType, RunnerName, cancellationToken);
    }
    catch (Exception ex) when (ex is not OperationCanceledException || !cancellationToken.IsCancellationRequested)
    {
      _logger.LogError(ex, "Error checking for next job: {Message}", ex.Message);
      throw new InvalidOperationException("Unable to check and start job.", ex);
    }
  }

  [MemberNotNull(nameof(_currentJob))]
  [MemberNotNull(nameof(_jobRepository))]
  [TsIgnore]
  private void CheckJobAndRepo()
  {
    CheckJobRepo();
    if (_currentJob == null)
    {
      throw new InvalidOperationException("Unable to set job items because no current job exists.");
    }
  }

  [MemberNotNull(nameof(_jobRepository))]
  [TsIgnore]
  private void CheckJobRepo()
  {
    if (_jobRepository == null)
    {
      throw new InvalidOperationException("Unable to set job items because no job repository is set.");
    }
  }

  [TsIgnore]
  private async Task AddInitialJob(JobWorkerSettings settings, CancellationToken cancellationToken)
  {
    CheckJobRepo();

    // Default to job should be due until 1 day from now
    var spanToCheck = TimeSpan.FromDays(1);
    if (settings.AddNextJobAfter.HasValue)
    {
      // If time between jobs is set, use this
      spanToCheck = settings.AddNextJobAfter.Value;
    }

    // calculate time until the next job should be in the database
    DateTime dueDate = DateTime.UtcNow.Add(spanToCheck);
    TParams? parameters = GetInitialJobParameters();

    // check if a job already exists that is due until the calculated date
    IJob<TParams, TResult>? existingJob = await _jobRepository.FindExistingJobAsync(
      settings.JobType,
      dueDate,
      parameters,
      true,
      cancellationToken);

    if (existingJob != null)
    {
      _logger.LogInformation(
        "Found existing job {ExistingJobId} due {DueDate}, skip adding initial job.",
        existingJob.Id,
        existingJob.DueDate);
    }
    else
    {
      // create a new job since none exist that is due until the calculated time
      await _jobRepository.AddJobAsync(settings.JobType, dueDate, parameters, cancellationToken);
    }
  }

  [TsIgnore]
  private async Task DeleteOldJobs(TimeSpan deleteOlderThan, string jobType, CancellationToken cancellationToken)
  {
    DateTime now = DateTime.UtcNow;

    // only delete once a day to prevent unneeded db calls on jobs with short intervals
    if (_lastDeletedAt.HasValue && now.Subtract(TimeSpan.FromHours(24)) < _lastDeletedAt.Value)
    {
      return;
    }

    CheckJobRepo();

    DateTime deleteBefore = now.Subtract(deleteOlderThan);
    int deleted = await _jobRepository.DeleteJobsAsync(jobType, JobState.Finished, deleteBefore, cancellationToken);
    _lastDeletedAt = DateTime.UtcNow;

    if (deleted > 0)
    {
      _logger.LogInformation(
        "Deleted {Deleted} jobs of type {JobType} that were finished before {DeleteBefore}.",
        deleted,
        jobType,
        deleteBefore);
    }
  }
}

File: src/NodeJS/JobWorkerBase.ts
import { add, addDays, sub } from "date-fns";
import { nanoid } from "nanoid";
import type { IJob } from "./Generated/IJob";
import type { IJobRepository } from "./Generated/IJobRepository";
import { JobState } from "./Generated/JobState";
import type { IJobWorkerBase } from "./Generated/Worker/IJobWorkerBase";
import { IJobWorkerMetrics } from "./Generated/Worker/IJobWorkerMetrics";
import type { IJobWorkerSettings } from "./Generated/Worker/Settings/IJobWorkerSettings";
import type { LoggerFactory, SimpleLogger } from "./Helper/LoggerFactory";
import { EmtpyLogger } from "./Helper/LoggerFactory";

/**
 * A base class that allows to implement a job worker.
 * @template TParams The type of the parameter a job has.
 * @template TResult The type of the result of a job.
 */
export abstract class JobWorkerBase<TParams, TResult> implements IJobWorkerBase<TParams, TResult> {
  private currentJob: IJob<TParams, TResult> | undefined;
  private baseLogger: SimpleLogger;
  private cancellationToken: AbortSignal;
  private executedJobs = 0;
  private lastJobDurationMs = 0;
  private lastJobFinishTime: Date | undefined;
  private lastJobDeleteTime: Date | undefined;

  /**
   * Initializes a new instance of the @see JobWorkerBase class.
   * @param {IJobWorkerSettings} settings The job worker settings.
   * @param {IJobRepository<TParams, TResult>} jobRepo An instance of the job repository.
   * @param {LoggerFactory} loggerFactory A factory function that creates an instance of a logger.
   */
  constructor(
    protected settings: IJobWorkerSettings,
    protected jobRepo: IJobRepository<TParams, TResult>,
    loggerFactory?: LoggerFactory,
  ) {
    if (!loggerFactory) {
      loggerFactory = (_, __): SimpleLogger => new EmtpyLogger();
    }
    this.baseLogger = loggerFactory(this.constructor, {
      module: "@droidsolutions-oss/job-service",
    });

    this.assertRepo(this.jobRepo);
  }

  /** The name of this worker as used in logs and written to the job. */
  public runnerName: Readonly<string>;

  /**
   * Returns metrics of this worker instance.
   * @returns {Record<string, unknown>} An object with executedJobs and lastJobDurationMs.
   */
  public getMetrics(): IJobWorkerMetrics {
    let totalSeconds = 0;

    if (this.settings.addNextJobAfter?.days) {
      totalSeconds += this.settings.addNextJobAfter.days * 24 * 60 * 60;
    }

    if (this.settings.addNextJobAfter?.hours) {
      totalSeconds += this.settings.addNextJobAfter.hours * 60 * 60;
    }

    if (this.settings.addNextJobAfter?.minutes) {
      totalSeconds += this.settings.addNextJobAfter.minutes * 60;
    }

    if (this.settings.addNextJobAfter?.seconds) {
      totalSeconds += this.settings.addNextJobAfter.seconds;
    }

    return {
      jobType: this.settings.jobType,
      executedJobs: this.executedJobs,
      lastJobDurationMs: this.lastJobDurationMs,
      lastJobFinishedAt: this.lastJobFinishTime,
      jobIntervallSeconds: this.settings.addNextJobAfter ? totalSeconds : undefined,
    };
  }

  /**
   * Should return a name that distinguishes this worker from other workers. Should be the same for all instances since an id is generated
   * an appended to the string given here.
   */
  public abstract getRunnerName(): string;

  /**
   * Should return the parameters for the initial job at startup. Only needed if an initial job should be created. Can return undefined if
   * no intiial job is needed or job has no parameters.
   */
  public abstract getInitialJobParameters(): TParams | undefined;

  /**
   * A hook that is called before a worker runs. Thisi s called before the job is fetched so it might be that no job is
   * available.
   */
  public abstract preJobRunHook(): void;

  /**
   * A hook that is called after the job is finished Is also called when no job was available.
   */
  public abstract postJobRunHook(): void;

  /**
   * Should contain the logic needed to handle the job. The promise should resolve to the result of the job.
   * @param job The job including its paramters.
   * @param cancellationToken A token to cancel the operation.
   * @returns {Promise<TResult>} A promise that contains the result of the job.
   */
  public abstract processJobAsync(job: IJob<TParams, TResult>, cancellationToken: AbortSignal): Promise<TResult>;

  /**
   * Starts the worker. After a configurable initial delay it will frequently poll for new jobs and take the one with
   * the oldest due date. Each runs calls @see preJobRunHook @see processJobAsync and @see postJobRunHook . If no job is
   * available @see processJobAsync is not called.
   * To stop the worker use the provided stoppingToken.
   * @param {AbortSignal} stoppingToken A token that can be used to stop the worker.
   */
  public async executeAsync(stoppingToken: AbortSignal): Promise<void> {
    this.cancellationToken = stoppingToken;
    await this.delay(this.settings.initialDelaySeconds, stoppingToken);

    const runnerId = nanoid(7);
    this.runnerName = `${this.getRunnerName()}-${runnerId}`;
    let firstRun = true;
    this.baseLogger.info(
      { runner: this.runnerName },
      "Starting runner %s watching for jobs with type %s with polling interval %ds.",
      this.runnerName,
      this.settings.jobType,
      this.settings.jobPollingIntervalSeconds,
    );

    while (!stoppingToken.aborted) {
      const executed = false;
      const startTime = process.hrtime.bigint();
      try {
        this.preJobRunHook();
        await this.handleJobRunAsync(this.settings, stoppingToken, firstRun);
        this.postJobRunHook();

        firstRun = false;

        stoppingToken.throwIfAborted();
        await this.delay(this.settings.jobPollingIntervalSeconds, stoppingToken);
      } catch (err) {
        if (stoppingToken.aborted) {
          this.baseLogger.info({ runner: this.runnerName }, "Stopped runner %s.", this.runnerName);
          break;
        }

        throw err;
      } finally {
        if (executed) {
          const endTime = process.hrtime.bigint();
          this.lastJobDurationMs = Number((endTime - startTime) / 1000000n);
        }

        this.lastJobFinishTime = new Date();
      }
    }
  }

  /**
   * A wrapper around the repository to help set the amount of items the job must process.
   * @param {number} items The amount of items of the job.
   */
  public async setTotalItemsAsync(items: number): Promise<void> {
    this.assertJob(this.currentJob);

    await this.jobRepo.setTotalItemsAsync(this.currentJob, items, this.cancellationToken);
  }

  /**
   * A wrapper around the repository to help add progress to the job.
   * @param {number} amount The amount of progress to add.
   */
  public async addProgressAsync(amount?: number): Promise<void> {
    this.assertJob(this.currentJob);

    await this.jobRepo.addProgressAsync(this.currentJob, amount, false, this.cancellationToken);
  }

  /**
   * A wrapper around the repository to help add failed progress to the job.
   * @param {number} amount The amount of failed progress to add.
   */
  public async addFailedProgressAsync(amount?: number): Promise<void> {
    this.assertJob(this.currentJob);

    await this.jobRepo.addProgressAsync(this.currentJob, amount, true, this.cancellationToken);
  }

  private assertRepo(repo: IJobRepository<TParams, TResult> | undefined): asserts repo {
    if (!repo) {
      throw new Error("Unable to perform action on job because job repository is not set.");
    }
  }

  private assertJob(job: IJob<TParams, TResult> | undefined): asserts job {
    if (!job) {
      throw new Error("Unable to perform action on job because no current job is set.");
    }
  }

  /**
   * A helper function that waits a given amount of time that can be cancelled with a token.
   * @param {number} seconds The amount of seconds to wait.
   * @param {AbortSignal} cancellationToken A token to cancel the waiting.
   * @returns {Promise<void>} A promise that resolves after the given amount of seconds or rejects when the token is cancelled.
   */
  private async delay(seconds: number, cancellationToken: AbortSignal): Promise<void> {
    return new Promise((resolve, reject) => {
      const abortHandler = (_event: Event) => {
        clearTimeout(timeout);
        reject(cancellationToken.reason);
      };

      cancellationToken.addEventListener("abort", abortHandler);
      const timeout = setTimeout(() => {
        cancellationToken.removeEventListener("abort", abortHandler);
        resolve();
      }, seconds * 1000);
    });
  }

  /**
   * Internal handler for job runs. Is called between pre and post hook for every run through the main loop.
   * @param {IJobWorkerSettings} settings The worker settings.
   * @param {AbortSignal} cancellationToken The cancellation token.
   * @param {boolean} firstRun If this is the first run.
   * @returns {Promise<boolean>} A promise that resolves to true when a job was executed or false if no job was found or it was resetted.
   */
  private async handleJobRunAsync(settings: IJobWorkerSettings, cancellationToken: AbortSignal, firstRun?: boolean): Promise<boolean> {
    let executed = false;
    this.currentJob = await this.getJobAsync(this.jobRepo, settings, cancellationToken);

    if (this.currentJob) {
      try {
        this.currentJob.result = await this.processJobAsync(this.currentJob, cancellationToken);
        await this.jobRepo.finishJobAsync(this.currentJob, settings.addNextJobAfter, cancellationToken);

        this.executedJobs++;
        executed = true;
      } catch (err) {
        this.baseLogger.warn(
          { err, jobId: this.currentJob.id, runner: this.runnerName },
          "Failed to process job %d, job is beeing resetted. Message: %s",
          this.currentJob.id,
          (err as Error).message,
        );

        await this.jobRepo.resetJobAsync(this.currentJob, cancellationToken);

        // re-throw if cancellation error so job loop ends
        if (cancellationToken.aborted) {
          throw err;
        }
      }
    } else if (firstRun && settings.addInitialJob) {
      try {
        await this.addInitialJob(settings, cancellationToken);
      } catch (err) {
        this.baseLogger.fatal({ err }, `Unable to create initial job: ${(err as Error).message}`);
        throw new Error("Unable to create initial job.");
      }
    }

    if (this.settings.deleteJobsOlderThan) {
      await this.deleteOldJobs(this.settings.jobType, this.settings.deleteJobsOlderThan, cancellationToken);
    }

    this.currentJob = undefined;

    return executed;
  }

  /**
   * Internal method to check for available jobs.
   * @param {IJobRepository<TParams, TResult>} repo The repository.
   * @param {IJobWorkerSettings} settings The settings.
   * @param {AbortSignal} cancellationToken The cancellation token.
   * @returns {Promise<IJob<TParams, TResult> | undefined>} The next due job or undefined if no job is found.
   */
  private async getJobAsync(
    repo: IJobRepository<TParams, TResult>,
    settings: IJobWorkerSettings,
    cancellationToken: AbortSignal,
  ): Promise<IJob<TParams, TResult> | undefined> {
    try {
      return await repo.getAndStartFirstPendingJobAsync(settings.jobType, this.runnerName, cancellationToken);
    } catch (err) {
      if (cancellationToken.aborted) {
        // app is probably shutting down, this is handled in the main loop
        throw err;
      }

      this.baseLogger.error({ err }, "Error checking for next job: %s", (err as Error).message);

      throw new Error(`Unable to start a new job: ${(err as Error).message}`);
    }
  }

  private async addInitialJob(settings: IJobWorkerSettings, cancellationToken: AbortSignal) {
    // calculate date until which a job with duedate should exist
    let dueDate = new Date();
    if (settings.addNextJobAfter) {
      // use intervall between jobs as limit
      dueDate = add(dueDate, settings.addNextJobAfter);
    } else {
      // Default to 1 day
      dueDate = addDays(dueDate, 1);
    }

    const params = this.getInitialJobParameters();
    const existingJob = await this.jobRepo.findExistingJobAsync(settings.jobType, dueDate, params, true, cancellationToken);

    if (existingJob) {
      this.baseLogger.info("Found existing job %d due %s, skipping add of initial job.", existingJob.id, existingJob.dueDate.toUTCString());
    } else {
      this.baseLogger.info("Did not find any existing %s job that is due until %s, adding initial job.", settings.jobType, dueDate);

      // create a new job if none exists that is due until the calculated time
      await this.jobRepo.addJobAsync(settings.jobType, dueDate, params, cancellationToken);
    }
  }

  private async deleteOldJobs(
    jobType: string,
    olderThan: { days?: number; hours?: number; minutes?: number; seconds?: number },
    cancellationToken: AbortSignal,
  ) {
    const current = new Date();

    // don't spam delete queries for jobs with short intervals
    if (this.lastJobDeleteTime && sub(current, { hours: 24 }) > this.lastJobDeleteTime) {
      return;
    }

    const deleteBefore = sub(current, olderThan);
    const deleted = await this.jobRepo.deleteJobsAsync(jobType, JobState.Finished, deleteBefore, cancellationToken);
    this.lastJobDeleteTime = current;

    if (deleted > 0) {
      this.baseLogger.info("Deleted %d old jobs of type %s.", deleted, jobType);
    }
  }
}
